components/style/gecko_conversions.rs

```diff
@@ -6,10 +6,50 @@
 //! Ideally, it would be in geckolib itself, but coherence
 //! forces us to keep the traits and implementations here
 
+#![allow(unsafe_code)]
+
 use app_units::Au;
+use gecko_bindings::bindings::{RawServoStyleSheetStrong, ServoComputedValuesStrong};
 use gecko_bindings::structs::nsStyleCoord_CalcValue;
+use gecko_bindings::sugar::refptr::{RefCounted, RawServoStyleSheetBorrowed, ServoComputedValuesBorrowed};
+use properties::ComputedValues;
+use std::mem::transmute;
+use std::ptr::null;
+use std::sync::Arc;
+use stylesheets::Stylesheet;
 use values::computed::{CalcLengthOrPercentage, LengthOrPercentage};
 
+macro_rules! unsafe_impl_refcounted {
+    ($strong:ident, $borrowed:ident, $servo:ident) => {
+        unsafe impl<'a> RefCounted<'a> for $servo {
+            type Strong = $strong;
+            type Borrowed = $borrowed<'a>;
+            fn into_strong(x: Arc<Self>) -> Self::Strong {
+                // we can't use a default method for this
+                // because transmute needs to know that the sizes are
+                // the same pre-monomorphization.
+                // We could use transmute_copy but that throws away
+                // the size check.
+                unsafe {transmute(x)}
+            }
+            fn null_strong() -> Self::Strong {
+                unsafe {transmute(null::<Self>())}
+            }
+            fn from_borrowed(mut x: Self::Borrowed) -> &'a Arc<Self> {
+                let arc = unsafe { &*(&mut x as *mut Self::Borrowed as *mut Arc<Self>) };
+                debug_assert!(&**arc as *const Self != null());
+                arc
+            }
+            fn into_borrowed(x: &'a Arc<Self>) -> Self::Borrowed {
+                unsafe { *(x as *const Arc<Self> as *const Self::Borrowed) }
```

Comment:
I thought that Self::Borrowed was actually a pointer to the box inside the Arc, not a pointer to the Arc itself. - Am I wrong?

If that is the case, then into_borrowed and from_borrowed need to be changed. Most notably, from_borrowed will probably have to take a closure which is called with the &'a Arc<Self> to ensure that it lives long enough, or the object returned will have to actually be some sort of RAII guard.